### ISSUE TITLE
Finish full migration to mimicked filesystem

### ISSUE THREAD

**Issue by eclipse-sailing-analytics-bot** [eclipse-sailing-analytics/sailing-analytics#6020]
| Field | Value |
|---|---|
| **Reporter** | @timtamtom123 |
| **Assignee** | @timtamtom123 |
| **Product** | Sailing Race Analytics |
| **Component** | Architecture |
| **Version** | unspecified |
| **Hardware** | All |
| **OS** | All |
| **Severity** | enhancement |
| **Priority** | P2 |
| **Created** | 2024-06-13T10:43:26Z |
| **Original Bug** | [Bug 6020](https://old-bugzilla.sapsailing.com/bugzilla/show_bug.cgi?id=6020) |

---

[bug 5938](https://github.com/eclipse-sailing-analytics/sailing-analytics/issues/5938) has been merged to main. A new sailing image has been created from scratch using the new setup script and we successfully launched a replica set. We have also ran two image upgrades for the mongo. The first to pull the latest git changes and the second to use the new version of the imagupgrade.sh. Both replicas launched from this image are in STARTUP2 and we are awaiting SECONDARY status.

To finish the migration we need to:
- Scale out the old mongo instances that used the old AMI (provided the new instances are healthy).
- Move all sailing images to the new image. First, once the new build is complete, we upgrade all the application replica sets. Then we need to change the image type. This can be done by clicking "move all application processes away from this replica sets master host to a new host". IMPORTANT: This needs to be done sooner rather than later in case an instance launched using the old image type, tries to deploy a process to a multi server on the new image, because it will try to call to /home/sailing/code/java/target/refreshInstance.sh. It may even be worth doing a manual image upgrade and installing a symbolic link at this path to /usr/local/bin/refreshInstance.sh just to be safe. 
For those that work based on auto replicas, such as "my". We click "move master process to a another instance". Then we click "update machine image for auto scaling replicas". And then scale accordingly to remove the old AMI and then bring in the new AMI.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

The MongoDB instances mongo0.internal.sapsailing.com and mongo1.internal.sapsailing.com have moved to the new image now, and the DNS records have been adjusted to their new internal IP addresses accordingly.

For the sailing instances, the following steps have so far been undertaken:
 - launching a new ARCHIVE candidate on the new 2.16 image
 - moving the first of three multi-servers to the 2.16 image (for the others I'm getting instance not found exceptions that I'll try to solve the same way that [bug 6023](https://github.com/eclipse-sailing-analytics/sailing-analytics/issues/6023) has solved it; build is underway and needs deployment to security-service then
 - moved security-service.sapsailing.com to a 2.16 image

Next up:
 - I'll retry moving the other two multi-instance nodes to 2.16 but may finally have to wait for a new build that has the exception handler
 - I'll start migrating my.sapsailing.com's master/replica now
 - I'll move all other replica sets to the new 2.16 image for auto-scaling

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

(In reply to Axel Uhl from comment 2 (on this issue))
> Next up:
>  - I'll retry moving the other two multi-instance nodes to 2.16 but may
> finally have to wait for a new build that has the exception handler

Release main-202406131557 ready and deployed to security-service. With this, I'm now starting to migrate the remaining two multi-servers.

>  - I'll start migrating my.sapsailing.com's master/replica now

Master migration triggered. After the master turned healthy for a short moment, the migration procedure triggered a re-start of the auto-replica's Java process in-place. This produced a healthy replica after some time, but still on the old image.

I then increased the number of desired auto-replicas from 1 to 2, producing another auto-replica, now from the new image. But there was a problem: the auto-scaler reduced the number of desired instances to 1 again after a few minutes based on the lack of traffic, and this terminated the only healthy replica. The new replica did not turn healthy as it couldn't connect to the now unhealthy master which is still hung up in tens of thousands of queued tasks, unable still to respond to replication requests.

This now leaves us with only an unhealthy but still re-calculating master, and one unhealthy auto-replica that cannot connect to master. :-(

The auto-replica will require manual attention once the master is healthy again.

>  - I'll move all other replica sets to the new 2.16 image for auto-scaling

Number two of three is currently being moved. Three will follow thereafter. The new ARCHIVE candidate is on its way but won't be ready for compareServers before tomorrow morning.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

(In reply to Axel Uhl from comment 3 (on this issue))

All production sailing servers are now running on the 2.16 image, and Hudson/DEV has been set up using the set-up script from a new AL2 instance (now on t3.medium, saving some cost). The failover archive is still on 2.15, but a new candidate is already launching on 2.16. Once this is ready and compares with no diffs, the next ARCHIVE rotation can be performed, and the last 2.15 imaged instance can be terminated.

Then, we can continue to work on [bug 6022](https://github.com/eclipse-sailing-analytics/sailing-analytics/issues/6022) ([Bug 6022](https://github.com/eclipse-sailing-analytics/sailing-analytics/issues/6022)), removing symbolic links that are no longer needed, with the exception of the central MongoDB instance which ideally should also receive an upgrade soon.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

Tested the MongoDB image upgrade successfully. The imageupgrade unit now produces valid logs, seen through journalctl -t imageupgrade.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

Sailing and Mongo Replica images now look good. To get them upgraded, I needed to manually intervene and remove the second and third argument from the build-crontab-and-cp-files invocation which exits with an error now if there is more than one positional argument.

The central MongoDB instance (dbserver.internal.sapsailing.com) no longer holds any reference to the Git workspace under /root/code. So we can now re-visit the links listed in [bug 6022](https://github.com/eclipse-sailing-analytics/sailing-analytics/issues/6022) and start removing stuff from configuration/ that is no longer needed, such as configuration/mongo_instance_setup. See commit 5ad607d78c619477602e863d39819cd7a0400f2c on main.
